src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs

```diff
@@ -88,6 +89,17 @@ protected void Publish()
                 return;
             }
 
+            // MeterListener has a static constructor that creates runtime metrics instruments.
+            // We need to ensure this static constructor is called before starting to publish the instrument.
+            // This is necessary because creating runtime metrics instruments will cause re-entry to the Publish method,
+            // potentially resulting in a deadlock due to the SyncObject lock.
+            // Sequence of the deadlock:
+            //   1. An application creates an early instrument (e.g., Counter) before the MeterListener static constructor is executed.
+            //   2. Instrument.Publish is called and enters the SyncObject lock.
+            //   3. Within the lock block, MeterListener is called, triggering its static constructor.
+            //   4. The static constructor creates runtime metrics instruments, causing re-entry to Instrument.Publish and leading to a deadlock.
```

Comment:
The ordinary locks like `lock (Instrument.SyncObject)` allow reentrancy. This should not lead to a deadlock. This description of the problem does not make sense to me.

Reply:
My guess is that the deadlock you have observed is A-B B-A dead lock between the `Instrument.SyncObject` and the static constructor lock. It is hard to tell for sure without seeing stacktraces of the deadlock. I believe these must be multiple threads involved. I do not think that this can deadlock when everything is executing on a single thread.

Reply:
Yeah it looked like inverted lock ordering between static constructor and sync object:
![deadlock](https://github.com/user-attachments/assets/fa90a3b2-aa7c-486f-b800-b0d4652c4de9)

I believe the fix addresses the issue but the comment description doesn't match.

Reply:
> My guess is that the deadlock you have observed is A-B B-A dead lock between the `Instrument.SyncObject` and the static constructor lock. It is hard to tell for sure without seeing stacktraces of the deadlock. I believe these must be multiple threads involved. I do not think that this can deadlock when everything is executing on a single thread.

I can share dump with you, but you are correct.

Reply:
> the comment description doesn't match.

Submitted #105539 to delete the wrong comment and simplify the fix.